### PR TITLE
Port `index.Tensor` to structured kernels.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #69607

Tracking issue: #55070

### CODE REVIEW

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (bdhirsh) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:480]: Not sure how important the extra overhead incurred from the dispatcher trip is here. We already know that we're in a QuantizedCPU kernel though, so maybe it's not too unreasonable to call `at::cpu::index`? If we do, we should rename the function `quantized_index` -> `quantized_cpu_index` to avoid confusion.

**Comment by bdhirsh** [comment]
I see a few CI failures. They look like they're mostly coming from unexpected successes in OpInfo tests though.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (bdhirsh) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:471]: Also, totally not needed but we could swap out the `torch::List` here with `iOptTensorRefList` while we're at it, to give unboxing quantized calls a small perf boost.

File: tools/codegen/api/types.py
Comment (bdhirsh) [inline tools/codegen/api/types.py:60]: this seems reasonable

**Comment by bdhirsh** [comment]
This looks good :). I'll approve when CI looks good

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (ysiraichi) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:480]: Ah, you are right. That sounds way more reasonable.

**Comment by ysiraichi** [comment]
There's a small problem with this PR: the implementation has its own dispatching criteria, which ignores the dispatcher.

**Why was it working before?**
CPU and CUDA were dispatched to the same function.

**Why doesn't it work anymore?**
CPU and CUDA are dispatched to different functions. Specifically, `set_output` behaves differently depending whether the operation was dispatched by the dispatcher to CPU or CUDA.

**How could we fix this?**
Having a way to check what dispatch key was used. This would allow us to either:
- `at::redispatch::index` to the right kernel
- copy everything to the corresponding device (this already happens with the `indices`, and continue the execution
 
@bdhirsh @ezyang 
What do you think?

**Comment by bdhirsh** [comment]
> There's a small problem with this PR: the implementation has its own dispatching criteria, which ignores the dispatcher.

@ysiraichi hmm maybe I'm just missing something. Where does this happen?

It looks like both before and after in this PR, CPU and CUDA both dispatch to the same kernel (`CPU, CUDA: index_out` in the yaml). So I'd expect the same meta function run, and I'd expect them to hit the same impl function.

Then the "extra dispatch" happens through `DispatchStub`: you call into `index_stub(device_type(), *this, sizes, strides);` in that shared cpu/cuda kernel, which ends up performing a (maybe slightly redundant) extra dispatch to the separate real kernels for cpu and cuda.

That seems like it was the case both before and after this PR though. The `DispatchStub` logic is also kinda necessary (even though it weirdly shares some of the responsibilities as the dispatcher), because it's responsible for handling cpu vector instruction code.

> Specifically, set_output behaves differently depending whether the operation was dispatched by the dispatcher to CPU or CUDA.

I don't think I'm following - don't we just end up invoking `TORCH_PRECOMPUTE_META_FUNC2(index, Tensor)` in both the CPU and CUDA case, which invokes the shared `build_index_op()` to do all of the `TensorIteratorBase` stuff that eventually calls `set_output()`?

**Comment by ysiraichi** [comment]
The problem here is exactly to which function `set_output` gets dispatched to. For example:

```python
>>> t = torch.rand(5)
>>> i = torch.tensor([1], device='cuda')
>>> t[i]
```

`t[i]` gets dispatched to the CUDA implementation. When `set_output(..., self.options())` gets called, it assumes it's in CUDA. But, `self.options().device()` gives CPU: assertion error!

**Comment by bdhirsh** [comment]
> The problem here is exactly to which function `set_output` gets dispatched to. For example:
> 
> ```python
> >>> t = torch.rand(5)
> >>> i = torch.tensor([1], device='cuda')
> >>> t[i]
> ```
> 
> `t[i]` gets dispatched to the CUDA implementation. When `set_output(..., self.options())` gets called, it assumes it's in CUDA. But, `self.options().device()` gives CPU: assertion error!

Ah, thanks for the example.

It looks like this happens because the logic for choosing with argument to use when setting the device guard is different between structured and unstructured. For example, before this PR this is was the cuda kernel for `index.Tensor` looked like:

```
at::Tensor wrapper_Tensor_index_Tensor(const at::Tensor & self, const c10::List<c10::optional<at::Tensor>> & indices) {
  const OptionalDeviceGuard device_guard(device_of(self));
  return at::native::index(self, indices);
}
```

The precedence is that we use the `self` argument if it exists - otherwise we pick the `out=` argument, and otherwise we look for other tensor-like args in the schema.

In structured kernel world though, we just use the device from whatever `TensorOptions` was passed into `TensorOptions`. So we effectively always use the device of the output.

The fact that we use a different argument to set device guards for structured vs unstructured seems like a bug worth looking into more.

@ngimel for this specific case though, what do you think of the current behavior of `index` for cuda? It definitely seems like an outlier - I don't know of any other ops that can take in mixed `cpu` and `cuda` tensor inputs, and choose to output a `cpu` tensor. Do you think that's the behavior we want and/or we're stuck with it anyway for BC reasons?

**Comment by ngimel** [comment]
It's definitely an outlier, and I think this behavior is encoded in python_variable_indexing.cpp and not in the aten proper. The use case for mixing (that people actually depend on) is using cpu indices for the cuda `self` tensor, but apparently the code got written in a way that allows other direction too.

**Comment by bdhirsh** [comment]
We talked about this offline a bit (@ngimel feel free to correct me if any of this sounds wrong):

(1) Since the `cpu_tensor[cuda_indices]` behavior wasn't really added intentionally, it might be worth trying to change `index` to error in that case explicitly and seeing if all tests pass. That could be an indicator of how BC-breaking it would actually be to remove that behavior, and then maybe we could just error when we see that behavior and not have to deal with the problem :)

(2) Otherwise, if we're stuck with this behavior, we agreed that `index` is a one-off special case that doesn't easily fit with structured kernels (because you can't reliably use the output tensor's device to set the device guard). And since there aren't really any other ops known to have the same behavior, it's probably not worth beefing up structured codegen / TIBase to support this specific case. So like we do for factory ops, we could just write an explicit meta kernel (at the cost of some code duplication).

(Maybe one other small factor @ysiraichi is that before, the plan for `IOptTensorRefList` was that it could only be used by structured kernels, as a way to make migration easier, meaning that unstructured ops that take in lists of tensors wouldn't enjoy the perf benefit that `IOptTensorRefList` gives you. But I guess that's not really an issue anymore if we go forward with the changes you have, where you ported all `TensorList` call-sites to use `ITensorList`.)

**Comment by ezyang** [comment]
`cpu_tensor[cuda_indices]` sounds crazy to me. We should drop it.

**Comment by ysiraichi** [comment]
Checking for `cpu_tensor[cuda_indices]` is fairly easy. That said, what about other devices? Should we just ignore them (and their combination as well) for now?

**Here's what I think:** ideally, this kernel should work on tensors on the same device only. That said, given that people depend on this behavior (i.e. CPU indices), I'd suggest the following:

- allow `<dev>_tensor[{cpu,dev}_indices]`
- disallow the rest (i.e. mixed non-CPU devices)

@bdhirsh @ngimel @ezyang 
What do you think?

**Comment by ezyang** [comment]
> Checking for cpu_tensor[cuda_indices] is fairly easy. That said, what about other devices? Should we just ignore them (and their combination as well) for now?

As Natalia points out, cuda_tensor[cpu_tensor], (when cpu tensor is a scalar, I think?), should work. There's actually a meta heuristic which works here: mixed cpu+cuda operations are OK when the cpu tensor in question is small. This is because you can directly ship the (small) cpu tensors contents to the CUDA kernel in the kernel launch parameters, without having to trigger an (expensive) HtD copy. So technically we could make cpu_tensor[cuda_tensor] work if cpu tensor was a scalar tensor and the output lived on CUDA but this would be *super* pointless so no point in allowing it haha.

**Comment by ngimel** [comment]
`cuda_tensor[cpu_tensor]` works not only for scalar cpu tensors, but also for (presumably) large index tensors. It stems from the history where we allowed indexing with lists (like numpy does), and lists are usually turned to cpu tensors. At this point downstream projects like torchvision depend on cuda_tensor[cpu_tensor] indexing, and they don't mind very much the ensuing copy.

**Comment by ysiraichi** [comment]
> So technically we could make `cpu_tensor[cuda_tensor]` work if cpu tensor was a scalar tensor and the output lived on CUDA but this would be super pointless so no point in allowing it haha.

Agreed.

> `cuda_tensor[cpu_tensor]` works not only for scalar cpu tensors, but also for (presumably) large index tensors.

Yes. After expanding the indices, we copy the tensors to `self` device.

-----

So, is [my proposal](https://github.com/pytorch/pytorch/pull/69607#issuecomment-1052076279) okay? Using that reasoning, we can port `index.Tensor` to structured kernels, while accommodating the pattern people depend on. That also enables us to easily (without any codegen changes) check for, possibly, ambiguous combination of mixed devices.

File: tools/autograd/templates/Functions.h
Comment (albanD) [inline tools/autograd/templates/Functions.h:38]: Why do you need to change this?

**Comment by ezyang** [comment]
Thumbs up for proposal for me.

**Comment by ngimel** [comment]
Thumbs up from me also

File: tools/autograd/templates/Functions.h
Comment (ysiraichi) [reply tools/autograd/templates/Functions.h:38]: **The problem**: in `IListIterator::operator*` I try to convert an `IValue` (boxed element) into an `OptionalTensorRef`. There, I'm mapping:

- `ivalue.isNone()` => `OptionalTensorRef{}`
- otherwise => implicitly conversion from `ivalue.toTensor()`

However, I can't create an `OptionalTensorRef` from an undefined tensor (there's a check for that).

------

While this change solve the issue above, I could, instead, change the mapping I do in `IList`:

- `ivalue.isNone() && !ivalue.toTensor().defined()` => `OptionalTensorRef{}`
- otherwise => implicitly conversion from `ivalue.toTensor()`

Do you think that's a better approach?

File: test/test_torch.py
Comment (ysiraichi) [inline test/test_torch.py:5465]: XLA tests seem to be failing here, and I'm not sure why. Is there any possibility that after calling `tensor.to('xla')`, they end up on CPU?

File: tools/autograd/templates/Functions.h
Comment (albanD) [reply tools/autograd/templates/Functions.h:38]: For your use case I am not sure. But this change does change the semantic of the current function.
I am not familiar about all the call site for this function but we should make sure that this change in semantic is ok there.

File: test/test_torch.py
Comment (ezyang) [reply test/test_torch.py:5465]: no, shouldn't be possible IMO

File: tools/autograd/templates/Functions.h
Comment (ysiraichi) [reply tools/autograd/templates/Functions.h:38]: You are right. I think it's best to handle this inside `IList`.

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34706811).

File: test/test_torch.py
Comment (ysiraichi) [reply test/test_torch.py:5465]: I think so.
Another thing I think could be happening is that when `x[:, ia, None, ib, 0]` is executed, there's some (maybe fallback) dispatch step in XLA that: (i) moves those tensors to CPU; and (ii) dispatches to the CPU kernel (probably not the problem here, though).
I remember that, sometimes, XLA fallbacks to CPU, right? Is this something that could be happening? One easy fix to this is by making this test `@onlyCUDA`. What do you think?

Comment (bdhirsh) [reply test/test_torch.py:5465]: XLA doesn't share error checking logic with our kernels, so their index kernel probably needs an update to check that base + indices are all on the same device.

That logic is [here](https://github.com/pytorch/xla/blob/8ffef19735f8fa3278f4d42c8d64e86c31af49dd/torch_xla/csrc/aten_xla_type.cpp#L1600) - I can make a patch for it

Comment (ysiraichi) [reply test/test_torch.py:5465]: Thanks for looking it up, Brian.
But, on second thoughts, I think that will only be a problem after XLA's signature also changes to the **new** one. Here's what I think is happening:

- `at::index` tries to dispatch to the XLA kernel
- There's no XLA kernel that matches the **new** signature (with `IOptTensorRefList`)
- It gets dispatched to the CPU kernel via calling `xla_cpu_fallback`

Let me know what you think.

Comment (ysiraichi) [reply test/test_torch.py:5465]: Nope. Forget about my last reply.
You were totally right, Brian. That's what's wrong with this CI failure. Could you make a patch for it?
What I said is real only for #73351.

Comment (bdhirsh) [reply test/test_torch.py:5465]: Patch is here: https://github.com/pytorch/xla/pull/3414. I'll be on PTO for the rest of this week, but I can help land the patch when I get back (unless we need to land sooner, then maybe @JackCaoG can help land).

File: aten/src/ATen/native/IndexingUtils.h
Comment (ezyang) [inline aten/src/ATen/native/IndexingUtils.h:24]: Index is not necessarily moved into the result, so it would be profitable to keep it a reference

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (ezyang) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:78]: for reference, in GitHub's review UI moving functions around like this makes reviewer's life harder as we have to manually verify that there were no substantive changes

Comment (ezyang) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:78]: Should this still be taking a TensorList?

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (ezyang) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:357]: Should info be saved in the struct for impl? We can do this with precompute right?

File: tools/autograd/templates/Functions.h
Comment (ezyang) [reply tools/autograd/templates/Functions.h:38]: This still needs updating right?

File: torch/nn/utils/rnn.py
Comment (ezyang) [inline torch/nn/utils/rnn.py:334]: Wow! This is surprising. Shouldn't unsorted_indices only be transferred to CPU if lengths is on CPU?

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (ysiraichi) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:357]: Hmm... Is it okay to use `AdvancedIndex` type (holds both a `Tensor` and a `std::vector<Tensor>`) with precompute? I thought precompute wasn't supposed to be used for `Tensor`.

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp
Comment (ezyang) [reply aten/src/ATen/native/TensorAdvancedIndexing.cpp:78]: > Is it better just to add the function signature there?

Well, there are a few ways you can do it. One is to forward declare. Another is to do the code motion in a separate PR stacked before this one.

**Comment by ysiraichi** [comment]
@ezyang @bdhirsh 
Apparently, some lazy tensor tests are failing. I've written about it below, so let me know what you think.

-----

### The Cause
- Call: `lazy_self[lazy_index]`
- Gets dispatched to the `CompositeExplicitAutograd` version of `index.Tensor`
- Gets dispatched to the `ltc_eager_fallback`, which copies the `lazy_self` tensor (only) to CPU
- Assert fail: can't have `lazy_self` on CPU and `lazy_indices` on Lazy device

-----

### Why does it fail?
- Only `lazy_self` gets copied to CPU
- Since `lazy_index` is passed as a list of **optional** tensors, it doesn't get copied to CPU
- `ts_eager_fallback` (called by `ltc_eager_fallback`) does not handle lists of optional tensors

-----

### Possible Solutions
1. Make `ts_eager_fallback` handle list of optional tensors (sounds better)
2. Remove `index.Tensor` testing for lazy tensors

**Comment by ezyang** [comment]
I think @bdhirsh has a PR that may help with this, I cc'ed you own it @ysiraichi

**Comment by ysiraichi** [comment]
My bad. I forgot to inspect the body of the constructor, too.
@ezyang Could you try again?

**Comment by ezyang** [comment]
I see XLA is still failing CI here, is there more follow up needed? Also, @ysiraichi, do you think maybe we can land the next two diffs in this stack first?

**Comment by JackCaoG** [comment]
Brian had a pr https://github.com/pytorch/xla/pull/3414 to fix the xla failure but I think it was not finished.

**Comment by ysiraichi** [comment]
@ezyang Yeah, sure. Do I need to do anything on my end?

**Comment by bdhirsh** [comment]
I just updated the XLA-side PR here (waiting for green CI) https://github.com/pytorch/xla/pull/3414

I also see a LazyTensor test failing - you probably need to update this test https://github.com/pytorch/pytorch/blob/3a8752db864022e785f4db6ad57988dff9977d54/test/cpp/lazy/test_lazy_ops.cpp#L4710 to also copy `indices` onto the lazy device (I'm guessing right now that test allows `indices` to be on cuda).

**Comment by ysiraichi** [comment]
> (I'm guessing right now that test allows indices to be on cuda).

That seems to be the case. But, one question: shouldn't the lazy tensor be turned to eager at that point?
After debugging a bit, it seems like `self` is on `lazy:0`. Should that be happening? Shouldn't `ts_eager_fallback` be turning that tensor into an eager one?
If that tensor was made eager, this problem wouldn't happen...

**Comment by bdhirsh** [comment]
> Shouldn't ts_eager_fallback be turning that tensor into an eager one?

Yeah, what you're saying sounds totally reasonable - it looks like today, the TS backend for LazyTensor doesn't have a lowering for `index.Tensor`, so I would expect it to hit the fallback. I'd probably need to pull the PR locally and fire up gdb to tell why.

Either way though, you'd still run into problems once LazyTensor eventually provides a lowering for `index.Tensor` (they would run the meta function from core, which would see `lazy_param` on `lazy:0`, and `indices` on `cuda`). So it still seems worth fixing regardless?

**Comment by ysiraichi** [comment]
> So it still seems worth fixing regardless?

I guess you are right.
But, if I understand it correctly, fixing that test would make it into another test, wouldn't it?

-----

#### `TestOneIndex` (Passing)
https://github.com/pytorch/pytorch/blob/3a8752db864022e785f4db6ad57988dff9977d54/test/cpp/lazy/test_lazy_ops.cpp#L4683-L4688

#### `TestOneIndexTransfer` (Failing)
https://github.com/pytorch/pytorch/blob/3a8752db864022e785f4db6ad57988dff9977d54/test/cpp/lazy/test_lazy_ops.cpp#L4708-L4711

**Comment by bdhirsh** [comment]
Sure - and you can update `TestOneIndexTransfer` to work by ensuring that `indices` is always allocated on cpu, I think? (Instead of the "default fallback device", which might be cuda

**Comment by ysiraichi** [comment]
Ah. I see what you mean.

**Comment by ysiraichi** [comment]
@bdhirsh I think there's only XLA and the BC CI failing. Let me know if there's anything missing here.

**Comment by bdhirsh** [comment]
@ysiraichi darn, do you mind rebasing? XLA CI is red because this PR is currently before the `tools/codegen` -> `torchgen` change, which is causing the XLA PR to complain

**Comment by ysiraichi** [comment]
@bdhirsh 
`linux-xenial-cuda11.3-py3.7-gcc7` seems to be failing because, for example, `TestMetaCUDA.test_dispatch_meta___getitem___cuda_bfloat16` is unexpectedly being successful at executing `aten.index.Tensor`.

I tried to investigate it for some time, but couldn't really solve it. I thought that removing `aten.index.Tensor` from `meta_dispatch_expected_failures` would solve the issue. However, it raises a `RuntimeError: failed to run: aten.index.Tensor`, which confused myself. Apparently, the error is due to `at::nonzero` not having a dispatch for Meta.

Could you help me on this one?

**Comment by ysiraichi** [comment]
Ah, never mind. I found the problem: `index.Tensor` calls `at::nonzero` (no Meta implementation) in the Meta function.

**Comment by bdhirsh** [comment]
I had a cuda build running overnight haha, glad you found it

**Comment by bdhirsh** [comment]
CI looks good minus XLA - just kicked off another round of the XLA CI PR here: https://github.com/pytorch/xla/pull/3414. When that's green I'll land both

**Comment by bdhirsh** [comment]
It was a bug in my XLA patch. Just pushed a fix, re-running.

**Comment by bdhirsh** [comment]
Apparently there's a new policy for merging changes that require XLA patches - we merge the XLA patch first, and in the core PR specify the commit hash from pytorch/XLA that our CI should run against.

Just updated this PR to point to the commit hash from https://github.com/pytorch/xla/pull/3414. Also fixed a lint error.

**Comment by bdhirsh** [comment]
Hmmm I was trying to update the PR myself to fix the lint, but I wasn't able to for some reason. @ysiraichi whenever you get around to it do you mind fixing the lint?

**Comment by ysiraichi** [comment]
@bdhirsh I think it's mergeable... Could you take a look?

**Review by bdhirsh — APPROVED** [review]
Thanks for all of the rounds of review!

**Comment by bdhirsh** [comment]
I'm going to force merge to unbreak XLA CI. Quick reasoning: before my last commit, I had the xla_commit_hash pointing at my xla-side patch (linked above), and all CI was green (except for the internal meta complaint, which I think should be safe to ignore since this PR doesn't actually require any internal changes). Now that the XLA PR has landed, I updated the xla_commit_hash to point at the updated xla tip-of-master commit

**Comment by ysiraichi** [comment]
@bdhirsh The problem was that I had forgotten to include `CUDAFunctions.h` when I was using `at::cuda::index_out`. Should be fixed.

**Comment by bdhirsh** [comment]
Thanks for the quick turnaround @ysiraichi.

The XLA failures also looked suspicious - I updated the xla_commit_hash to point to the new XLA PR here https://github.com/pytorch/xla/pull/3623, and I also added a TORCH_PIN on that PR pointed back to this one, so we can confirm CI is green on both PR's.

**Comment by ysiraichi** [comment]
@bdhirsh Apparently, the XLA CI failure is not related to this PR (as it was already failing from `1b4b5acafd`). The other one seems to be a rebase thing.

**Comment by ysiraichi** [comment]
3 of the CI failures seem unrelated. They are error-ing with the following (the file path changes, though):

```
Can't use 'tar -xzf' extract archive file: /home/runner/actions-runner/_work/_actions/_temp_bce906eb-67d7-4b84-bf85-551083ebbd41/86b214bd-c446-4aa1-9254-44e5ea2ba9b7.tar.gz
```

The mobile CI is failing due to a `ConnectionResetError`. Finally, XLA is failing because:

```
> git checkout 30a1adfabe0945409447ccc2b6dc9947ee1c98bf
fatal: reference is not a tree: 30a1adfabe0945409447ccc2b6dc9947ee1c98bf
```

This was probably related to the change you made before. Should I not have rebased this PR?
@bdhirsh Should I try rebasing it?

**Comment by bdhirsh** [comment]
I think it's safe to ignore the non-XLA failures - they were all green before we merged last time.

For XLA, let me poke around. Rebasing won't help because I updated the xla_commit_hash to point to the XLA-side change. I'm going to try updating it.

(The XLA-side patch is also taking a while because I'm having trouble getting a local XLA build working to debug the errors with :( I think I know what the fix is though, testing it out)

**Comment by malfet** [comment]
> @pytorchbot revert -m “Sorry this PR broke MPS tests” -c nosignal

@janeyx99 your quotes are too fancy for mergebot ;) (it understand only ansi ones, and you are using nice unicodes ones)

**Comment by malfet** [comment]
So, at least on M1 machines, after this PR, import triggers following warning:
```
>>> import torch
/Users/malfet/miniconda3/envs/test-mps/lib/python3.8/site-packages/torch/library.py:66: UserWarning: Overriding a previously registered kernel for the same operator and the same dispatch key
  operator: aten::index.Tensor(Tensor self, Tensor?[] indices) -> Tensor
    registered at aten/src/ATen/RegisterSchema.cpp:6
  dispatch key: Meta
  previous kernel: registered at aten/src/ATen/RegisterCPU.cpp:39683
       new kernel: registered at /dev/null:219 (Triggered internally at  ../aten/src/ATen/core/dispatch/OperatorEntry.cpp:133.)
  self.m.impl(name, dispatch_key, fn)
```

File: torch/_meta_registrations.py
Comment (bdhirsh) [inline torch/_meta_registrations.py:175]: @ezyang let me know if you want to keep this around. It seems useful to have a python impl, but we can't register it now that `index.Tensor` is structured without some other hacks (we'll end up with a duplicate registration)

Comment (ezyang) [reply torch/_meta_registrations.py:175]: Yeah just keep it without registering

**Comment by bdhirsh** [comment]
This diff requires a hot fix in fbcode, but I wasn't able to import it. Hot fix is here, I'll rebase it on top of the internal diff once it enters the diff train: D37015324

**Comment by bdhirsh** [comment]
Hmm @janeyx99 do you know why the mergebot is complaining?

It looks like the PR actually merged successfully - I see this commit on master, which points back to this PR: https://github.com/pytorch/pytorch/commit/7b3a0ff87a91c98574ddfd98532062d78c49b2c7.

If you think it's unreliable then I'll ignore it.

**Comment by janeyx99** [comment]
> Hmm @janeyx99 do you know why the mergebot is complaining?
> 
> It looks like the PR actually merged successfully - I see this commit on master, which points back to this PR: [7b3a0ff](https://github.com/pytorch/pytorch/commit/7b3a0ff87a91c98574ddfd98532062d78c49b2c7).
> 
> If you think it's unreliable then I'll ignore it.

Oh weird--I do think this is not reliable and you can ignore it--the issue tracking it is https://github.com/pytorch/test-infra/issues/397

**Pull request closed without merging**
